### PR TITLE
KOGITO-3716 Automatic close/promote mvn stage

### PR DESCRIPTION
**Pull request by radtriste**
https://issues.redhat.com/browse/KOGITO-3716

Dependent PRs:
- https://github.com/kiegroup/kogito-pipelines/pull/91
- https://github.com/kiegroup/kogito-runtimes/pull/865
- https://github.com/kiegroup/optaplanner/pull/1010
- https://github.com/kiegroup/kogito-examples/pull/435

### CODE REVIEW

File: Jenkinsfile.release [hunk 1]
```diff
@@ -61,18 +66,25 @@ pipeline {
...
+        BOT_AUTHOR='radtriste-bot-account'
+        BOT_AUTHOR_CREDS_ID='radtriste-bot' // username/password creds
+        GITHUB_TOKEN_CREDS_ID='radtriste-gh-token' // secret text creds
+
+        // Maven Nexus release parameters
+        DEFAULT_STAGING_REPOSITORY = 'https://repository.jboss.org/nexus/content/groups/kogito-public/'
+        NEXUS_RELEASE_URL = 'https://repository.stage.jboss.org/nexus'
+        NEXUS_RELEASE_REPOSITORY_ID = 'jboss-releases-repository'
+        NEXUS_STAGING_PROFILE_ID = '2161b7b8da0080'
+        NEXUS_BUILD_PROMOTION_PROFILE_ID = '1966c60aff37d'
```
Comment (radtriste) [inline Jenkinsfile.release]: will be set back after testing

**Review by Kevin-Mok — APPROVED** [review]

**Comment by Kevin-Mok** [comment]
When I try to run `mvn clean test` in `jenkins/tests`, I'm getting the following error:
```
[INFO] Scanning for projects...
[INFO] [INFO] -------------------< org.kie.jenkins:jenkins-tests >--------------------
[INFO] Building Kogito :: Jenkins Spock Tests 1.0.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[WARNING] The POM for 
org.kie:jenkins-pipeline-shared-libraries:jar:1.0.1 is 
missing, no dependency information available
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  1.138 s
[INFO] Finished at: 2020-12-04T19:49:34-05:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal on project jenkins-tests: 
Could not resolve dependencies for project 
org.kie.jenkins:jenkins-tests:jar:1.0.0-SNAPSHOT: Failure to 
find org.kie:jenkins-pipeline-shared-libraries:jar:1.0.1 in 
https://repo.jenkins-ci.org/releases was cached in the local 
repository, resolution will not be reattempted until the 
update interval of jenkins-releases has elapsed or updates 
are forced -> [Help 1]
[ERROR] [ERROR] To see the full stack trace of the errors, 
re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full 
debug logging.
[ERROR] [ERROR] For more information about the errors and 
possible solutions, please read the following articles:
[ERROR] [Help 1] 
http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException
```

**Comment by radtriste** [comment]
yeah it is because the shared-libraries jar is not deployed to Nexus (yet).
That is why I am installing it into the GHAction: https://github.com/kiegroup/kogito-pipelines/pull/91/files#diff-b59e021276476287d3bd9c994f5a4f0ddefbfa2989aab669a5db711bedaa9601R32
You will need, for now, to checkout https://github.com/kiegroup/jenkins-pipeline-shared-libraries and `mvn install` to run the tests.
It is because we import now `MavenCommand` and so for testing the pipeline, we need it into the classpath

**Comment by radtriste** [comment]
Updated pom.xml to get shared library deployed via https://github.com/kiegroup/jenkins-pipeline-shared-libraries/pull/126

File: Jenkinsfile.release [hunk 2]
```diff
@@ -581,6 +554,17 @@ void addSkipTestsParam(buildParams){
     addBooleanParam(buildParams, 'SKIP_TESTS', params.SKIP_TESTS)
 }
 
+void addMavenSettingsParam(buildParams){
+    addStringParam(buildParams, 'MAVEN_SETTINGS_CONFIG_FILE_ID', params.MAVEN_SETTINGS_CONFIG_FILE_ID)
+}
+
+void addMavenNexusReleaseParam(buildParams){
```
Comment (rsynek) [inline Jenkinsfile.release]: Maybe use a plural to signal there is more than a single property?
```suggestion
void addMavenNexusReleaseParams(buildParams){
```

File: Jenkinsfile.release [hunk 3]
```diff
@@ -322,22 +300,21 @@ pipeline {
...
                     buildJob(OPERATOR_DEPLOY, buildParams)
                 }
             }
         }
 
-        stage('Is staging repository released?') {
+        stage('Are staged artifacts released?') {
             when {
-                expression { return !isCreateReleaseBranchesOnly() && !getArtifactsRepositoryParam() } // Execute only if artifacts repository was not given, which means the staging repository has been created
+                expression { return !isCreateReleaseBranchesOnly() && isArtifactsDeploy() && !getArtifactsRepositoryParam() } // Execute only if artifacts repository was not given, which means the staging repository has been created
```
Comment (rsynek) [inline Jenkinsfile.release]: Just nitpicking to make the comment fit the screen.
```suggestion
                // Execute only if artifacts repository was not given, which means the staging repository has been created.
                expression { return !isCreateReleaseBranchesOnly() && isArtifactsDeploy() && !getArtifactsRepositoryParam() } 
```

File: Jenkinsfile.release [hunk 4]
```diff
@@ -291,7 +269,7 @@ pipeline {
                     addImageBuildParams(buildParams, '', getKogitoImagesTempTag(), true, false)
                     addStringParam(buildParams, 'EXAMPLES_URI', env.EXAMPLES_URI)
                     addStringParam(buildParams, 'EXAMPLES_REF', env.EXAMPLES_REF)
-                    addStringParam(buildParams, 'MAVEN_ARTIFACT_REPOSITORY', env.STAGING_REPOSITORY) // Get artifacts from staging repository
+                    addStringParam(buildParams, 'MAVEN_ARTIFACT_REPOSITORY', getArtifactsRepositoryParam() ?: env.DEFAULT_STAGING_REPOSITORY) // Get artifacts from staging repository
```
Comment (rsynek) [inline Jenkinsfile.release]: Maybe wrap the ternary operator in a "smart" getter to avoid duplication?

File: jenkins/tests/pom.xml [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+
+        <logback.configration>logback-test.xml</logback.configration>
+        <logdir>${project.build.directory}/log</logdir>
+        <test.loglevel>ERROR</test.loglevel>
+        <log.logback.version>1.2.3</log.logback.version>
+        <log.slf4j.version>1.7.25</log.slf4j.version>
+        <build-helper-maven-plugin.version>1.9.1</build-helper-maven-plugin.version>
+
+        <maven.compiler.source>1.6</maven.compiler.source>
+        <maven.compiler.target>1.6</maven.compiler.target>
```
Comment (rsynek) [inline jenkins/tests/pom.xml]: Is this version required by the test tools?

File: jenkins/tests/pom.xml [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+        <junit.version>4.12</junit.version>
+        <junit.plugin.version>1.24</junit.plugin.version>
+        <surefire.pluginVersion>2.22.0</surefire.pluginVersion>
+
+        <logback.configration>logback-test.xml</logback.configration>
+        <logdir>${project.build.directory}/log</logdir>
+        <test.loglevel>ERROR</test.loglevel>
+        <log.logback.version>1.2.3</log.logback.version>
+        <log.slf4j.version>1.7.25</log.slf4j.version>
+        <build-helper-maven-plugin.version>1.9.1</build-helper-maven-plugin.version>
```
Comment (rsynek) [inline jenkins/tests/pom.xml:56]: Would be nice to keep the same naming convention as we already have e.g. here: https://github.com/kiegroup/kogito-runtimes/blob/master/kogito-build-parent/pom.xml#L85-L95

Dependencies:
`version.{groupId}.{optional artifactId}`
Plugins:
`version.{name}.plugin`

**Review by rsynek — COMMENTED** [review]
Nice work. A couple of minor improvements suggested.

File: jenkins/tests/pom.xml (hunk 5 above)
Comment (radtriste) [reply jenkins/tests/pom.xml]: removed

File: jenkins/tests/pom.xml (hunk 6 above)
Comment (radtriste) [reply jenkins/tests/pom.xml:56]: will address this in https://issues.redhat.com/browse/BXMSPROD-1125 with solution of parent pom

**Comment by radtriste** [comment]
@rsynek addressed your comments in my latest commit

**Review by rsynek — APPROVED** [review]

**Pull request merged**